src/org/sosy_lab/common/NativeLibraries.java

```diff
@@ -95,65 +98,67 @@ public static OS guessOperatingSystem() {
       }
 
-      @Var String prop = System.getProperty("os.arch.data.model");
-      if (isNullOrEmpty(prop)) {
-        prop = System.getProperty("sun.arch.data.model");
+      Architecture osArch = guessOsArchitecture();
+      if (osArch == X86_64 && isVm32Bit()) {
+        currentArch = X86;
+      } else {
+        currentArch = osArch;
       }
 
-      if (!isNullOrEmpty(prop)) {
-        switch (prop) {
-          case "32":
-            currentArch = Architecture.X86;
-            break;
-          case "64":
-            currentArch = Architecture.X86_64;
-            break;
-          default:
-            throw new UnsatisfiedLinkError(
-                "Unknown value for os.arch.data.model: '"
-                    + prop
-                    + "', please report this together with information about your system "
-                    + "(OS, architecture, JVM).");
-        }
-      } else {
+      return currentArch;
+    }
 
-        prop = StandardSystemProperty.JAVA_VM_NAME.value();
-        if (!isNullOrEmpty(prop)) {
-          prop = Ascii.toLowerCase(prop);
-
-          if (prop.contains("32-bit") || prop.contains("32bit") || prop.contains("i386")) {
-
-            currentArch = Architecture.X86;
-          } else if (prop.contains("64-bit")
-              || prop.contains("64bit")
-              || prop.contains("x64")
-              || prop.contains("x86_64")
-              || prop.contains("amd64")) {
-
-            currentArch = Architecture.X86_64;
-          } else {
-            throw new UnsatisfiedLinkError(
-                "Unknown value for java.vm.name: '"
-                    + prop
-                    + "', please report this together with information about your system "
-                    + "(OS, architecture, JVM).");
-          }
-        } else {
-          throw new UnsatisfiedLinkError("Could not detect system architecture");
-        }
+    private static Architecture guessOsArchitecture() {
+      @Var String osArch = StandardSystemProperty.OS_ARCH.value();
+      if (isNullOrEmpty(osArch)) {
+        throw new UnsatisfiedLinkError("No value for os.arch" + REPORT_MESSAGE);
       }
 
-      return currentArch;
+      osArch = Ascii.toLowerCase(osArch.replace(" ", ""));
+
+      switch (osArch) {
+        case "i386":
+        case "i686":
+        case "x86":
+          return Architecture.X86;
+        case "amd64":
+        case "x86_64":
+          return Architecture.X86_64;
+        case "aarch64":
+          return Architecture.ARM64;
+        default:
+          throw new UnsatisfiedLinkError(
+              "Unknown value for os.arch: '"
+                  + StandardSystemProperty.OS_ARCH.value()
+                  + "'"
+                  + REPORT_MESSAGE);
+      }
+    }
+
+    /** Check whether the JVM is executing in 32 bit version. */
+    private static boolean isVm32Bit() {
+      return nullToEmpty(StandardSystemProperty.JAVA_VM_NAME.value()).matches(".*(32-?bit|i386).*")
```

Comment:
This does not have the same semantics as the previous check because it case sensitive. Is this intended?

Reply:
thanks for the finding. fixed.